interactive bots: Create tic-tac-toe bot by maydhak.

maydhak opened the pull request:
Wrote code and docs for bot that plays tic-tac-toe with the user.
GCI task link: https://codein.withgoogle.com/dashboard/task-instances/6115034390331392/

smarx commented:
*Automated message from Dropbox CLA bot*

@maydhak, it looks like you've already signed the Dropbox CLA. Thanks!

maydhak commented:
I just saw the Travis errors from the first commit, so I've fixed them now. I committed the changes and rebased so that it was all in one commit. Hopefully it's fine now.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+                ]
+
+    positions = "Here are the coordinates of each position. (Parentheses and spaces are optional.) \n" \
+                "(1, 1)  (1, 2)  (1, 3) \n(2, 1)  (2, 2)  (2, 3) \n(3, 1) (3, 2) (3, 3) \n " \
```
I guess you need to specify which item within a coordinate represents a column or row.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        for space in row:
+            row_string += space.strip() + " "
+        row_string += "]\n"
+        return row_string
```
More Pythonic:
```
row_string = ' '.join([e.strip() for e in row])
return '[{}]'.format(row_string)
```

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        boardstring = ""
+        for row in board:
+            boardstring += output_mode(self.display_row(row), mode)
+        return boardstring
```
Same for this:
```Python
return ''.join([output_mode(self.display_row(r), mode) for r in board])
```

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        full = False
+        board_state = ""
+        for row in board:
+            for space in row:
```
Instead of `space`, how about `element`?

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+            full = True
+        return full
+
+    def win_conditions_2(self, board, triplets):
```
Maybe just `win_conditions`?

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        won = False
+        for set in triplets:
+            if board[set[0][0]][set[0][1]] == board[set[1][0]][set[1][1]] == board[set[2][0]][set[2][1]] != "_":
+                won = True
```
You can simply return from here.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+
+    def win_conditions_2(self, board, triplets):
+        won = False
+        for set in triplets:
```
`set` is used by Python as well. I think you can just use `triplet`.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+    def win_conditions_2(self, board, triplets):
+        won = False
+        for set in triplets:
+            if board[set[0][0]][set[0][1]] == board[set[1][0]][set[1][1]] == board[set[2][0]][set[2][1]] != "_":
```
Can you add a comment explaining the win condition?

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+                            "Here are the coordinates of each position. (Parentheses and spaces are optional). \n" \
+                            "(1, 1)  (1, 2)  (1, 3) \n(2, 1)  (2, 2)  (2, 3) \n(3, 1) (3, 2) (3, 3) \n"
+
+    def __init__(self, aboard):
```
You can `board` here as well.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        ''' Gets the locations of the board that are still blank. Used when the computer makes its move. '''
+        blank_locations = []
+
+        for row in range(0, 3):
```
Simply `range(3)` works.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+
+        return blank_locations
+
+    def get_x_locations(self, board):
```
`get_x_locations` and `get_blank_locations` have some code duplication.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        blanks_list = []
+        if o_found:
+            for position in triplet:
+                if board[position[0]][position[1]] == "_":
```
This pattern is being repeated a lot. Maybe create a function which takes position and returns value?
```Python
def get_value(board, position):
    return board[position[0]][position[1]]
```

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+
+    def computer_move(self, board):
+        ''' The computer's logic for making its move. '''
+        myBoard = copy.deepcopy(board)  # First the board is copied; used later on
```
`my_board`

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        ''' The computer's logic for making its move. '''
+        myBoard = copy.deepcopy(board)  # First the board is copied; used later on
+        blank_locations = self.get_blank_locations(myBoard)
+        xs = self.get_x_locations(board)  # Gets the locations that already have x's
```
`x_locations` will make it more readable.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+                row = location[0]
+                col = location[1]
+                board[row][col] = "o"
+                return board
```
You can move `board` outside 'if...else`.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+                board[row][col] = "o"
+                return board
+
+        else:  # This logic is used on all other moves.
```
No need for `else` as you are return from `if`.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+        else:  # This logic is used on all other moves.
+            # First I'll check if the computer can win in the next move. If so, that's where the computer will play.
+            # The check is done by replacing the blank locations with o's and seeing if the computer would win in each case.
+            for location in blank_locations:
```
`for row, col in blank_locations:`

umairwaheed replied:
Same pattern can be applied below.

umairwaheed commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,345 @@
+                row = location[0]
+                col = location[1]
+                myBoard[row][col] = "x"
+                if self.win_conditions_2(myBoard, self.triplets) == True:
```
Simply `self.win_conditions_2(myBoard, self.triplets)`

umairwaheed commented:
@maydhak, pretty impressive. Some feedback left.

maydhak commented:
I've updated the PR to address your comments.

showell commented:
How hard would it be to switch from coordinates to just labeling the squares 1 through 9?  Both in terms of the UI and in terms of your internal representation?

~~~
1  2  3
4  5  6
7  8  9

showell commented:
For the UI piece of accepting moves, you may want to support human language:

upper left corner
top edge
upper right corner
left edge
middle (or center)
right edge
bottom left corner
bottom edge
bottom right corner

showell commented:
From an architectural standpoint, you try to create strong separation and boundaries between these concepts:

* game coordination (starting the game, ending the game, giving the human help)
* translating human input to commands your game engine understands
* talking back to the human (the computation piece should just be game_won=True, then another layer should be "good move, you win!")
* internal game representation (with no AI or UI, just keep track of the moves)
* displaying the board
* AI: how to have the computer calculate the next move

If you carefully separate these components, most of them should be easy to test in isolation.  Particularly for the AI piece, you should be able to write automated tests that feed your AI engine and verify that they make the right moves for maybe five to ten typical game scenarios.

This is an ambitious project, but we will merge this, I promise!  Let us know how much you want to work on this.  We can merge a working version and then continue to refine it as needed.  Also, feel free to ask any mentor to approve your task at this point, as what you have so far is very impressive and a lot of work.  If you get the task approved, just be sure to comment back here about where things stand.

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+
+# -------------------------------------
+class TicTacToeGame(object):
+    smarter = True
```
It would be more user friendly if the user didn't had to change the source code to change this setting. Have you considered using of configuration file?

maydhak replied:
Thanks for the feedback! I didn't/don't intend to make it a setting that could be changed. If the user wants to change smarter because the game is too difficult for them, I might add something in the future so that they can change smarter via a command to the bot. But right now, I think the game has the right degree of difficulty, and there's no need to change this.

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+# -------------------------------------
+class TicTacToeGame(object):
+    smarter = True
+    # If smarter is True, the computer will do some extra thinking when it makes its move - it'll be harder for the user.
```
This line seems to be too long.

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+        in the triplet are blank. '''
+        won = False
+        for triplet in triplets:
+            if self.get_value(board, triplet[0]) == self.get_value(board, triplet[1]) == self.get_value(board, triplet[2]) != "_":
```
This line seems to be too long.

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+
+        # Assuming nobody will win in their next move, now I'll find the best place for the computer to win.
+        for row, col in blank_locations:
+            if 'x' not in my_board[row] and my_board[0][col] != 'x' and my_board[1][col] != 'x' and my_board[2][col] != 'x':
```
This line seems to be too long.

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+    flat = sum(board, [])
+    return True if flat == flatInitial else False
+
+class tictactoeHandler(object):
```
Conventionally class names should be in camel case:

https://www.python.org/dev/peps/pep-0008/#class-names

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+flatInitial = sum(initial_board, [])
+def first_time(board):
+    flat = sum(board, [])
+    return True if flat == flatInitial else False
```
How about `return flat == flatInitial`?

femnad commented on contrib_bots/lib/tictactoe-bot.py:
```diff
@@ -0,0 +1,329 @@
+        return return_string
+
+# -------------------------------------
+flatInitial = sum(initial_board, [])
```
Conventionally Python variables are named with lowercase words separated by underscores:

https://www.python.org/dev/peps/pep-0008/#id45

maydhak commented:
Hi @femnad and @showell!
Thanks for all of your feedback. I've updated the PR with changes for all but one of those requested by @femnad, and I replied to that above. I've asked for the task to be approved on GCI. However, I definitely appreciate all of the helpful suggestions given by @showell as well, and would like to incorporate them in my code in the next few weeks, but post-GCI. I won't have much time over the next couple of weeks as I've got something important coming up (a personal commitment). I really do want to create a polished version of this bot.

showell commented:
@maydhak Sounds good!  Let me know if you are ready to merge this version as is.  I'm also happy to wait for you to polish it some more before merging.

maydhak commented:
Sure, please merge now. When I update my code with the improvements in a couple of weeks, should I push to the same branch (so updates will come on this pull request) or do that on a new branch and make a new PR?

showell commented:
Once we merge this, this PR will clear, so you can either re-use your branch or start a new branch.

If you have a moment, is there any chance you can re-push with a tiny change to force a new Travis build?  The latest build here was happening during a time when we had a Travis flake.  You can do this:

```
git checkout <your-branch>
git commit --amend
# don't edit anything, just save
git push origin +<your-branch

maydhak commented:
I've re-pushed.

showell commented:
Thanks, @maydhak!  There are now Travis errors that I think are actually about your code formatting.

maydhak commented:
@showell Updated, Travis is passing now :)

showell commented:
Merged!  Thanks @maydhak!  I look forward to seeing new stuff in a couple weeks.  Thanks for all your hard work during GCI.
